File: src/instruction.ts
import { BorshSchema, borshSerialize, borshDeserialize, Unit } from 'borsher';

// Определение enum типов
enum InstructionType {
    OldBuy = 1,
    OldWithdraw,
    ObtainTicket,
    Buy = 4,
    Withdraw = 5,
    UpdateState = 252,
    Migrate = 253,
    AdminWithdraw = 254,
    Initialize = 255,
}

export class Initialize {
    static readonly SCHEMA = BorshSchema.Struct({
        instruction: BorshSchema.u8,
        lootboxId: BorshSchema.u16,
        vaultBump: BorshSchema.u8,
        stateBump: BorshSchema.u8,
        maxSupply: BorshSchema.u32,
        beginTs: BorshSchema.u32,
        endTs: BorshSchema.u32,
        signer: BorshSchema.Array(BorshSchema.u8, 33),
        name: BorshSchema.String,
        prices: BorshSchema.Vec(BorshSchema.u64),
        baseUrl: BorshSchema.String,
    });

    instruction: InstructionType = InstructionType.Initialize;
    lootboxId: number;
    vaultBump: number;
    stateBump: number;
    maxSupply: number;
    beginTs: number;
    endTs: number;
    signer: Uint8Array;
    name: string;
    prices: number[];
    baseUrl: string;

    constructor(lootboxId: number, vaultBump: number, stateBump: number, maxSupply: number, beginTs: number, endTs: number, signer: Uint8Array, name: string, prices: number[], baseUrl: string) {
        this.lootboxId = lootboxId;
        this.vaultBump = vaultBump;
        this.stateBump = stateBump;
        this.maxSupply = maxSupply;
        this.beginTs = beginTs;
        this.endTs = endTs;
        this.signer = signer;
        this.name = name;
        this.prices = prices;
        this.baseUrl = baseUrl;
    }
}

export class Migrate {
    static readonly SCHEMA = BorshSchema.Struct({
        instruction: BorshSchema.u8,
        stateBump: BorshSchema.u8,
    });

    instruction: InstructionType = InstructionType.Migrate;
    stateBump: number

    constructor(stateBump: number) {
        this.stateBump = stateBump;
    }
}

export class UpdateState {
    static readonly SCHEMA = BorshSchema.Struct({
        instruction: BorshSchema.u8,
        lootboxId: BorshSchema.u16,
        stateBump: BorshSchema.u8,
        maxSupply: BorshSchema.u32,
    });

    instruction: InstructionType = InstructionType.UpdateState;
    lootboxId: number;
    stateBump: number;
    maxSupply: number;

    constructor(lootboxId: number, stateBump: number, maxSupply: number) {
        this.lootboxId = lootboxId;
        this.stateBump = stateBump;
        this.maxSupply = maxSupply;
    }
}

export class Buy {
    static readonly SCHEMA = BorshSchema.Struct({
        instruction: BorshSchema.u8,
        lootboxId: BorshSchema.u16,
        ticketBumps: BorshSchema.Vec(BorshSchema.u8),
        ticketSeed: BorshSchema.u32,
    });

    instruction: InstructionType = InstructionType.Buy;
    lootboxId: number;
    ticketBumps: Array<number>;
    ticketSeed: number;


    constructor(lootboxId: number, ticketBumps: Array<number>, ticketSeed: number) {
        this.lootboxId = lootboxId;
        this.ticketBumps = ticketBumps;
        this.ticketSeed = ticketSeed;
    }
}


export class Signature {
    static readonly SCHEMA = BorshSchema.Struct({
        recId: BorshSchema.u8,
        rs: BorshSchema.Array(BorshSchema.u8, 64),
    });

    recId: number;
    rs: Uint8Array;

    constructor(recId: number, rs: Uint8Array) {
        this.recId = recId;
        this.rs = rs;
    }
}

export class ObtainTicket {
    static readonly SCHEMA = BorshSchema.Struct({
        instruction: BorshSchema.u8,
        lootboxId: BorshSchema.u16,
        ticketBump: BorshSchema.u8,
        ticketId: BorshSchema.u32,
        expireAt: BorshSchema.u32,
        signature: Signature.SCHEMA,
    });

    instruction: InstructionType = InstructionType.ObtainTicket;
    lootboxId: number;
    ticketBump: number;
    ticketId: number;
    expireAt: number;
    signature: Signature;

    constructor(lootboxId: number, ticketBump: number, ticketId: number, expireAt: number, signature: Signature) {
        this.lootboxId = lootboxId;
        this.ticketBump = ticketBump;
        this.ticketId = ticketId;
        this.expireAt = expireAt;
        this.signature = signature;
    }
}

export class Withdraw {
    static readonly SCHEMA = BorshSchema.Struct({
        instruction: BorshSchema.u8,
        lootboxId: BorshSchema.u16,
        expireAt: BorshSchema.u32,
        signature: Signature.SCHEMA,
        tickets: BorshSchema.u8,
        amounts: BorshSchema.Vec(BorshSchema.u64),
    });

    instruction: InstructionType = InstructionType.Withdraw;
    lootboxId: number;
    tickets: number;
    amounts: number[];
    expireAt: number;
    signature: Signature;

    constructor(lootboxId: number, tickets: number, amounts: number[], expireAt: number, signature: Signature) {
        this.lootboxId = lootboxId;
        this.tickets = tickets;
        this.amounts = amounts;
        this.expireAt = expireAt;
        this.signature = signature;
    }
}

export class AdminWithdraw {
    static readonly SCHEMA = BorshSchema.Struct({
        instruction: BorshSchema.u8,
        lootboxId: BorshSchema.u16,
        amount: BorshSchema.u64,
    });

    instruction: InstructionType = InstructionType.AdminWithdraw;
    lootboxId: number;
    amount: number;

    constructor(lootboxId: number, amount: number) {
        this.lootboxId = lootboxId;
        this.amount = amount;
    }
}

export function serializeInitialize(instruction: Initialize): Uint8Array {
    return borshSerialize(Initialize.SCHEMA, instruction);
}

export function serializeBuy(instruction: Buy): Buffer {
    return borshSerialize(Buy.SCHEMA, instruction);
}

export function serializeObtainTicket(instruction: ObtainTicket): Uint8Array {
    return borshSerialize(ObtainTicket.SCHEMA, instruction);
}

export function serializeWithdraw(instruction: Withdraw): Uint8Array {
    return borshSerialize(Withdraw.SCHEMA, instruction);
}

export function serializeMigrate(instruction: Migrate): Uint8Array {
    return borshSerialize(Migrate.SCHEMA, instruction);
}

export function serializeUpdateState(instruction: UpdateState): Uint8Array {
    return borshSerialize(UpdateState.SCHEMA, instruction);
}

export function serializeAdminWithdraw(instruction: AdminWithdraw): Uint8Array {
    return borshSerialize(AdminWithdraw.SCHEMA, instruction);
}

File: program/src/instruction.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::program_error::ProgramError;
use solana_program::pubkey::Pubkey;

#[repr(u8)]
#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
#[borsh(use_discriminant = true)]
pub enum Instruction {
    Withdraw(WithdrawParam) = 2,
    ObtainTicket(ObtainTicketParams) = 3,
    Buy(BuyParam) = 4,
    UpdateState(UpdateStateParams) = 252,
    MigrateToV3(MigrateToV3Params) = 253,
    AdminWithdraw {
        lootbox_id: u16,
        amount: u64
    } = 254,
    Initialize(InitializeParams) = 255,
}
#[derive(Clone, PartialEq, BorshSerialize, BorshDeserialize, Debug)]
pub struct BuyParam {
    pub lootbox_id: u16,
    pub ticket_bumps: Vec<u8>,
    pub ticket_seed: u32,
}
#[derive(Clone, PartialEq, BorshSerialize, BorshDeserialize, Debug)]
pub struct InitializeParams {
    pub lootbox_id: u16,
    pub vault_bump: u8,
    pub state_bump: u8,
    pub max_supply: u32,
    pub begin_ts: u32,
    pub end_ts: u32,
    pub signer: [u8; 33],
    pub name: String,
    pub prices: Vec<u64>,
    pub base_url: String,
}

#[derive(Clone, PartialEq, BorshSerialize, BorshDeserialize, Debug)]
pub struct MigrateToV3Params {
    pub state_bump: u8,
}

#[derive(Clone, PartialEq, BorshSerialize, BorshDeserialize, Debug)]
pub struct UpdateStateParams {
    pub state_bump: u8,
    pub lootbox_id: u16,
    pub enabled_fields: u32,
    pub max_supply: u32,
    pub begin_ts: u32,
    pub end_ts: u32,
}

impl UpdateStateParams {
    const MAX_SUPPLY: u32 = 1;
    const BEGIN_TS: u32 = 2;
    const END_TS: u32 = 4;

    fn is_field(&self, flag: u32) -> bool {
        (self.enabled_fields & flag) == flag
    }

    pub fn is_max_supply(&self) -> bool {
        self.is_field(Self::MAX_SUPPLY)
    }

    pub fn is_begin_ts(&self) -> bool {
        self.is_field(Self::BEGIN_TS)
    }

    pub fn is_end_ts(&self) -> bool {
        self.is_field(Self::END_TS)
    }
}

#[derive(Clone, PartialEq, BorshSerialize, BorshDeserialize, Debug)]
pub struct ObtainTicketParams {
    pub lootbox_id: u16,
    pub bump: u8,
    pub id: u32,
    pub expire_at: u32,
    pub signature: Signature,
}

#[derive(Clone, PartialEq, BorshSerialize, BorshDeserialize, Debug)]
pub struct TransferParams {
    pub amount: u64
}

#[derive(Clone, PartialEq, BorshSerialize, BorshDeserialize, Debug)]
pub struct Signature {
    pub rec_id: u8,
    pub rs: [u8; 64],
}

#[derive(Clone, PartialEq, BorshSerialize, BorshDeserialize, Debug)]
pub struct WithdrawParam {
    pub lootbox_id: u16,
    pub expire_at: u32,
    pub signature: Signature,
    pub tickets: u8,
    pub amounts: Vec<u64>, // 1 for NFT
}

impl Instruction {
    pub fn unpack(input: &[u8]) -> Result<Self, ProgramError> {
        let instruction = Instruction::try_from_slice(input)?;

        Ok(instruction)
    }

    pub fn name(&self) -> &'static str {
        match self {
            Instruction::Buy(_) => "Buy",
            Instruction::Withdraw(_) => "Withdraw",
            Instruction::ObtainTicket(_) => "ObtainTicket",
            Instruction::MigrateToV3(_) => "MigrationToV3",
            Instruction::AdminWithdraw { .. } => "AdminWithdraw",
            Instruction::Initialize(_) => "Initialize",
            Instruction::UpdateState(_) => "UpdateState",
        }
    }
}